### PR TITLE
New agents

### PR DESCRIPTION
**Pull request by m-wojnar**
Zmiany:

- dodałem agenta UCB,
- dodałem agenta GradientBandit (opis w książce RL Intro by Sutton & Barto, 37-40),
- w e-greedy dodałem wykładnicze zanikanie wartości Q, co pomaga w scenariuszach niestacjonarnych,
- dodałem prosty reward w rozszerzeniu dla 802.11ax jako `rate * n_successful / n_all`,
- dodałem w dokumentacji nowych agentów i wprowadziłem małe poprawki w dokumentacji custom agents, aby kod zgadzał się z poprawkami wprowadzonymi w e-greedy.

### CODE REVIEW

File: reinforced_lib/agents/e_greedy.py [hunk 1]
```diff
@@ -39,32 +37,40 @@ class EGreedy(BaseAgent):
...
 
-    def __init__(self, n_arms: jnp.int32, e: jnp.float32, optimistic_start: jnp.float32 = 0.0) -> None:
+    def __init__(
+            self,
+            n_arms: jnp.int32,
+            e: Scalar,
+            optimistic_start: Scalar = 0.0,
+            alpha: Scalar = 0.0
+    ) -> None:
+        assert 0 <= alpha <= 1
```
Comment (Wotaker) [inline reinforced_lib/agents/e_greedy.py:52]: Podoba mi się tutaj i w pozostałych agentach dodanie tych assertów na parametry!

File: reinforced_lib/exts/ieee_802_11_ax.py [hunk 2]
```diff
@@ -85,6 +85,13 @@ def success_probability(self, snr: float, *args, **kwargs) -> np.ndarray:
     def action(self, mcs: int, *args, **kwargs) -> int:
         return mcs
 
+    @observation(observation_type=gym.spaces.Box(-np.inf, np.inf, (1,)))
+    def reward(self, mcs: int, n_successful: int, n_failed: int, *args, **kwargs) -> float:
```
Comment (Wotaker) [inline reinforced_lib/exts/ieee_802_11_ax.py:89]: O tym już rozmawialiśmy, wydaje się rozsądny taki reward, w miarę możliwości tych prostych agentów.

File: reinforced_lib/rlib.py [hunk 3]
```diff
@@ -352,7 +355,7 @@ def sample(
         elif isinstance(sample_observations, tuple):
             state, action = self._agent.sample(state, sample_key, *sample_observations)
         else:
-            state, action = self._agent(state, sample_key, sample_observations)
+            state, action = self._agent.sample(state, sample_key, sample_observations)
```
Comment (Wotaker) [inline reinforced_lib/rlib.py:358]: O, dobrze że to zauważyłeś! Był potencjał na niezłego buga...

File: reinforced_lib/agents/ucb.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        Quality values for each arm
+    N : array_like
+        Number of tries for each arm
+    """
+
+    Q: Array
+    N: Array
+
+
+class UCB(BaseAgent):
```
Comment (Wotaker) [inline reinforced_lib/agents/ucb.py:29]: Wygląda dobrze. przydało by się testy dopisać, al e to można z tego taska zrobić.

File: reinforced_lib/agents/gradient_bandit.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        """
+
+        mask = jnp.ones_like(state.H, dtype=jnp.bool_).at[action].set(False)
+        pi = jax.nn.softmax(state.H)
+
+        R = jnp.where(state.n == 1, reward, state.R)
+
+        H = state.H
+        H = jnp.where(mask, H, H + lr * (reward - R) * (1 - pi))
+        H = jnp.where(mask, H - lr * (reward - R) * pi, H)
```
Comment (Wotaker) [inline reinforced_lib/agents/gradient_bandit.py]: dlaczego nie połączyłeś?
```
H = jnp.where(mask, H - lr * (reward - R) * pi, H + lr * (reward - R) * (1 - pi))
```

File: reinforced_lib/agents/gradient_bandit.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    n : int
+        Number of the step.
+    """
+
+    H: Array
+    R: Scalar
+    n: jnp.int64
+
+
+class GradientBandit(BaseAgent):
```
Comment (Wotaker) [inline reinforced_lib/agents/gradient_bandit.py:33]: Tutaj też należało by w przyszłości jakiś test dopisać

**Review by Wotaker — APPROVED** [review]
Jakbyś mi tylko odpowiedział na te pytanie w _gradient_bandit.py_; ogólnie approve!

File: reinforced_lib/rlib.py (hunk 3 above)
Comment (m-wojnar) [reply reinforced_lib/rlib.py:358]: To prawda xd

File: reinforced_lib/agents/ucb.py (hunk 4 above)
Comment (m-wojnar) [reply reinforced_lib/agents/ucb.py:29]: Zgadzam się, ogólnie dobrze byłoby napisać testy do wielu części obecnego kodu, ale stwierdziłem, że w tej chwili postaram się zrobić w miarę działającą wersję, a jak nam wystarczy czasu to zrobimy taski z pisania testów i przykładów.

File: reinforced_lib/agents/gradient_bandit.py (hunk 6 above)
Comment (m-wojnar) [reply reinforced_lib/agents/gradient_bandit.py:33]: Tak jest, podobnie jak wyżej pomyślałem, że zrobimy z tego później osobne taski

File: reinforced_lib/agents/gradient_bandit.py (hunk 5 above)
Comment (m-wojnar) [reply reinforced_lib/agents/gradient_bandit.py]: Nie zauważyłem, za bardzo skupiłem się na matematycznym wzorze ;) To można zapisać nawet jeszcze ładniej:
`H = H + lr * (reward - R) * jnp.where(mask, -pi, 1 - pi)`.

Comment (m-wojnar) [reply reinforced_lib/agents/gradient_bandit.py]: Ogólnie teraz widzę, że cały update da się dużo prościej zrobić - wrzuciłem poprawkę.
